### PR TITLE
Resume offload panics before stop teardown

### LINKED ISSUE #263: Pending offload panic is not resumed on the stop path: on_stop runs and decorators observe Ok on a doomed incarnation
Found in the 2026-08-16 review round (multi-agent sweep plus two external reviews, each claim adjudicated case-by-case). Empirically reproduced; repro below.

## The bug

A pending offload panic is never resumed on the stop path, so the incarnation completes its entire orderly-stop protocol before the panic fires:

- An offloaded future that unwinds is caught and stored in the incarnation's panic slot (`raw.rs:1555-1557` → `RawDisposal::record`, `raw.rs:288-301`).
- The only receive-side resume is in `next_ready` (`raw.rs:1603-1610`). `recv`'s two stop branches (`raw.rs:1418`, `raw.rs:1430`) call `freeze_and_report()` and return `None` without consulting the slot, and `try_recv`'s draining branch (`raw.rs:1453-1460`) bypasses `next_ready` entirely.
- `Handler::run` (`actor.rs:366-397`) then drains the frozen prefix through `handle`, runs `on_stop`, and returns `Ok(())`.
- The incarnation runner re-raises the stored panic only after `run` returns (`raw.rs:2027-2049`), so a raw decorator wrapping `Handler` observes `Ok(())` for an incarnation that then panics underneath it.

Observed with the repro:

```
on_stop_ran   = true
drained_msg2  = true          <- a further handler delivery ran after the panic was pending
decorator_saw = Some("Ok(())")
exits         = ["decorated: Panicked { message: Some(\"owned offload panic\") }"]
```

The *published* exit is still `Panicked` with the right message, so supervision and restart classification are unaffected. The break is the crash-only contract: user teardown code (commits, acks, "clean stop" side effects in `on_stop` or in decorator code after `inner.run()`) executes on a failed incarnation.

## Spec tension

- SPEC §5.5 (~SPEC.md:1129-1130): panics in the offloaded future "resume on the actor task, so supervision classifies them as an ordinary actor panic."
- SPEC §5.2 (~SPEC.md:993-999): `on_stop` "does not run when … a handler — live or draining — returned `Err` or panicked (the incarnation is failed; cleanup is the crash-only path, `Drop`)."

The exclusion list says "a *handler* … panicked", so the offload case falls in a literal gap — but §5.5's framing says it should be covered. Needs a one-line adjudication either way; I read §5.5 as saying the exclusion applies.

## Fix shape

Consult the panic slot in `recv`'s two stop branches and in `try_recv`'s `is_stopping` branch: resume before returning `None`. The unwind then lands through `Handler::run`, which yields exactly §5.2's stated behavior (no `on_stop`, no further deliveries, decorator sees the unwind). The existing test `queued_offload_panic_survives_stop_and_handler_error` (`tests/offloads.rs:1294-1327`) cannot see the difference — its actor has a no-op `on_stop` — so the regression test must probe `on_stop`/decorator observables from the test body (the PR #249 discipline).

## Repro

Out-of-tree binary against the facade (public API only):

```rust
use std::sync::{
    Arc,
    atomic::{AtomicBool, Ordering},
};
use std::time::Duration;

use shelterwood::{
    Actor, Context, ExitError, ExitKind, ExitResult, Handler, LifecycleEventKind, LifecycleItem,
    Mailbox, RawActor, RawContext, RawOnceDef, Readiness, StopContext, Tree,
};

struct Flags {
    on_stop_ran: AtomicBool,
    decorator_saw: std::sync::Mutex<Option<String>>,
    drained: AtomicBool,
}

struct Inner {
    flags: Arc<Flags>,
}

impl Actor for Inner {
    type Msg = u32;
    type Args = Arc<Flags>;

    async fn init(args: Self::Args, _: &mut Context<'_, Self>) -> Result<Self, ExitError> {
        Ok(Self { flags: args })
    }

    async fn handle(&mut self, message: Self::Msg, context: &mut Context<'_, Self>) -> ExitResult {
        if message == 1 {
            let guard = context
                .offload_scoped(
                    async {
                        panic!("owned offload panic");
                    },
                    |_| 99u32,
                    Duration::MAX,
                )
                .expect("offload accepted");
            guard.finished().await;
            // The offload panic is now recorded in the incarnation's panic slot.
            context.stop();
            Ok(())
        } else {
            // Any later delivery proves the drain kept running after the
            // pending offload panic.
            self.flags.drained.store(true, Ordering::SeqCst);
            Ok(())
        }
    }

    async fn on_stop(&mut self, _: &mut StopContext<'_, Self>) {
        self.flags.on_stop_ran.store(true, Ordering::SeqCst);
    }
}

struct Decorator {
    inner: Handler<Inner>,
    flags: Arc<Flags>,
}

impl RawActor for Decorator {
    type Msg = u32;

    fn readiness() -> Readiness {
        Readiness::Manual
    }

    async fn run(&mut self, raw: &mut RawContext<Self::Msg>) -> ExitResult {
        let result = self.inner.run(raw).await;
        let seen = match &result {
            Ok(()) => "Ok(())".to_owned(),
            Err(error) => format!("Err({error:?})"),
        };
        *self.flags.decorator_saw.lock().expect("probe mutex") = Some(seen);
        result
    }
}

#[tokio::main(flavor = "multi_thread", worker_threads = 4)]
async fn main() {
    let flags = Arc::new(Flags {
        on_stop_ran: AtomicBool::new(false),
        decorator_saw: std::sync::Mutex::new(None),
        drained: AtomicBool::new(false),
    });
    let mut tree = Tree::new();
    let actor = tree
        .add_raw_once(
            "decorated",
            RawOnceDef::new(Decorator {
                inner: Handler::new(Arc::clone(&flags)),
                flags: Arc::clone(&flags),
            })
            .readiness(Readiness::Manual)
            .expect("valid readiness"),
        )
        .expect("valid actor");
    let system = tree.spawn().expect("runtime is available");
    let mut events = system.scope().subscribe_lifecycle();
    actor.send(1).await.expect("mailbox accepts before readiness");
    // A second message sits in the accepted prefix so the drain has work.
    actor.send(2).await.expect("mailbox accepts before readiness");

    let started = system.wait_started().await;
    println!("wait_started: {started:?}");

    let _ = system.shutdown(Duration::from_secs(2)).await;

    let mut exits = Vec::new();
    while let Some(item) = events.recv().await {
        if let LifecycleItem::Event(event) = item
            && let LifecycleEventKind::Exited { id, exit, .. } = &event.kind
        {
            exits.push(format!("{}: {:?}", id.as_str(), exit.kind()));
        }
    }

    println!("on_stop_ran   = {}", flags.on_stop_ran.load(Ordering::SeqCst));
    println!("drained_msg2  = {}", flags.drained.load(Ordering::SeqCst));
    println!(
        "decorator_saw = {:?}",
        flags.decorator_saw.lock().expect("probe mutex")
    );
    println!("exits         = {exits:?}");
    let _ = ExitKind::Completed;
}
```

## Acceptance criteria

- [ ] Adjudicate SPEC §5.2's exclusion list: does "panicked" cover offload panics? Amend §5.2 or §5.5 so the two agree.
- [ ] `recv`'s stop branches and `try_recv`'s draining branch resume a pending offload panic before yielding `None`.
- [ ] Regression test: an offload panic followed by `stop()` skips `on_stop`, delivers no further messages, and a wrapping decorator does not observe `Ok` — all judged from the test body.

### PR DESCRIPTION
**Pull request by ralexstokes**
## Summary

- resume a retained offload panic before either `recv` stop branch can return `None`
- resume the same panic before `try_recv` starts draining a frozen accepted prefix
- pin the stop-path contract through user-visible handler/decorator observations
- adjudicate SPEC §5.2 so incarnation-owned offload panics are failed incarnations that skip delivery and `on_stop`

## Root cause

Offload panics were retained in the incarnation resources, but the receive-side resume lived only in the normal `next_ready` path. Stop and drain branches bypassed it, allowing further handler work and clean-stop hooks to run before the outer incarnation runner finally re-raised the panic.

## Impact

An offload panic pending when stop begins now unwinds through the handler composition point immediately. No later accepted message is delivered, `on_stop` does not run, and a raw decorator cannot observe a false `Ok(())`.

## Validation

- focused `recv`/decorator and stopping-`try_recv` regressions
- full offload suite: 27/27 passed
- `./tools/dev just ci` — 658 workspace tests plus Clippy, doctests, rustdoc, API reachability, formatting, and packaged-doc checks

Closes #263

### CODE REVIEW

**Comment by ralexstokes** [comment]
Adversarial review completed on head `a8a0618fd7d999d69a3ad2f63c7ece2327b05407`.

### Areas attacked

- Issue #263 acceptance criteria and the SPEC §5.2/§5.5 adjudication.
- `PanicSlot::{record,take}`, `RawResources::resume_pending_panic`, and destructive ownership of the retained payload.
- Both `RawContext::recv` stop branches (local self-stop after the driver handshake and externally cancelled shutdown) plus the stopping `try_recv` drain branch.
- Freeze/cancel/join ordering, the raw-incarnation primary/cleanup panic precedence, and the possibility of a retained payload being resumed twice.
- `Handler::run` delivery/drain/`on_stop` sequencing and unwind propagation through a raw decorator.
- The framework lock rule: the panic-slot mutex is released by `take()` before panic resumption; no wake, user drop, formatting, callback, or unwind resumption was introduced under that mutex.
- Regression determinism and whether the tests could pass for the wrong reason.

### Finding and direct fix

I found no remaining production-code defect, but did confirm a meaningful test-observability gap in the submitted regression coverage:

- The drain/decorator test did not independently pin the local-stop `recv` check. If that `resume_pending_panic()` call was removed, Handler's newly fixed `try_recv` drain path still resumed the panic and the test stayed green.
- No submitted test entered the externally cancelled `recv` branch, so removing that branch's new check also left the suite green.
- The direct `try_recv` fixture had no accepted frozen-prefix message, weakening its claim that the panic wins before drain.

I fixed those findings directly in commit `a8a0618`:

- added a `MailboxShutdown::Discard` variant of the Handler/decorator test, bypassing `try_recv` so the local-stop `recv` branch is independently required;
- added a deterministic external-shutdown raw actor that queues the offload panic, waits for its shutdown token, and then enters `recv`, pinning the external branch and the original panic diagnostic;
- gated the stopping `try_recv` actor until a mailbox message is accepted, so the test now exercises a real frozen prefix.

Mutation verification was explicit: after temporarily removing both `recv` resume calls, the discard/local test failed because `on_stop` ran, and the external test failed with its post-`recv` unreachable diagnostic instead of the retained offload panic. Both passed after restoring the implementation.

### Verdict

No unresolved correctness, lock-rule, double-resume, or teardown-ordering findings remain at this head. The destructive slot take gives a single resume; the outer incarnation boundary retains actor-panic precedence while cleanup remains secondary; and each stop bypass is now independently observable.

### Validation

- focused branch regressions after restoration: passed
- complete `offloads` integration suite: **29/29 passed**
- `./tools/dev just ci`: **660/660 tests passed**, plus Clippy with `-D warnings`, doctests, rustdoc, public API reachability, formatting, packaged-doc checks, and Nix formatting
- `git diff --check`: clean

**Comment by ralexstokes** [comment]
## Independent review — Codex's P2 **CONFIRMED** (with a corrected remedy), plus one further P2

Reviewed head `a8a0618`. Second-opinion pass, independent of the authoring agent's self-review.

### 1. Codex `raw.rs:1437` — CONFIRMED (P2)

The escape hatch I went looking for — "the offload is polled by the actor task, so the interleaving is unreachable" — **does not exist**. The offload runs on its own runtime task:

- `raw.rs:1569` — `let task = runtime::spawn_actor_work(SharedOffloadFuture(Arc::clone(&state)));`
- `shelterwood-runtime/src/spawn.rs:124-131` — `spawn_actor_work` is `tokio::spawn`.

And the stop path **requests, never joins**: `RawResources::freeze` (`raw.rs:1027-1043`) → `OffloadResource::cancel` (`raw.rs:971-985`) fires the latch, calls `SharedOffloadState::cancel`, then `task.abort()`. `cancel` (`raw.rs:910-922`) sets `cancelled = true`, cannot take the future while `polling == true`, and fires `finished` anyway — no handshake with the poller. The only real join is `join_offloads` (`raw.rs:1069-1087`), reached solely from the epilogue at `raw.rs:2035`, i.e. **after** `RawActor::run` has returned.

The precise window Codex names is real: in `SharedOffloadFuture::poll` the Err arm calls `finish_poll` (clearing `polling`, releasing the mutex) at `raw.rs:953` *before* `self.0.record(payload)` at `raw.rs:954`.

**Reproduced on this branch** — offloaded future panics, a `Drop` inside it blocks so the payload is unwinding but unrecorded; `Drain` handler wrapped in a raw decorator:

```
REPRO on_stop_ran      = true
REPRO decorator_result = Some("ok")
REPRO exits            = ["in-flight-panic: Panicked { message: Some(\"in-flight offload panic\") }"]
```

That is #263's original symptom table, on the fixed branch.

**Two corrections to Codex's framing:**

1. **Severity is P2, not a blocker.** This is not a regression (main behaves identically here), and the published exit is still `Panicked`, so supervision/restart classification is unaffected. What the repro falsifies is the *sentence this PR newly added to the SPEC*, not the fix.
2. **The prescribed remedy is not implementable as stated.** "The stop transition must synchronize with in-progress polls" cannot be done at that site: `try_recv` is synchronous and cannot join an offload task, and hoisting `join_offloads` ahead of the drain would change stop ordering for every actor. The right resolution is to state the guarantee **conditionally**.

**Suggested fix — `SPEC.md:999-1001`.** The current text overclaims:

> "An offload panic therefore fails the incarnation before any further delivery or `on_stop`."

The preceding clause ("when an incarnation-owned offload panic *resumes at a receive boundary*") is accurate; the "therefore" sentence is not entailed by it. Replace with a retention-conditioned form plus an explicit non-guarantee:

> An offload panic **already retained when a receive boundary is reached** therefore fails the incarnation before any further delivery or `on_stop`. As with `Guard::finished()` (§5.5) this is not a join: a panic that lands after the last receive boundary is still the incarnation's exit, but cannot suppress `on_stop`.

§5.5 (`SPEC.md:1123-1125`) already gets this right for the neighbouring case — *"They are not a join guarantee."* The happens-before that makes the conditional version genuinely deliverable is real and is exactly what all four new tests exercise: `disposal.record` (`raw.rs:1560`) precedes `finished.fire()` (`raw.rs:949`), so an ordinary-completion `Guard::finished().await` guarantees a populated slot. Worth naming that cancellation also fires `finished` (`raw.rs:921`) *without* recording.

### 2. Second finding — P2: the change silently widens to **disposal** panics, not just offload panics

`PanicSlot` is a single shared slot for every `RawDisposal::record` (`raw.rs:288-300`). `freeze_and_report()` — called immediately before each new `resume_pending_panic()` — can itself deposit a **user destructor** panic into that slot: discarded continuations (`raw.rs:1038`), armed timers (`1039`), queued offload events (`1041`), and offload futures via `dispose` (`904-908`). Those now unwind out of `recv`/`try_recv` as the incarnation's **primary** panic instead of being collected as `cleanup` in the epilogue (`raw.rs:2031-2033`).

Reproduced with `continue_with` on a message whose `Drop` panics, then `ctx.stop()`:

| | `on_stop_ran` | `decorator_result` | exit |
|---|---|---|---|
| PR head | `false` | `None` | `Panicked("discarded continuation destructor panic")` |
| all three new lines removed (= main) | `true` | `Some("ok")` | same |

The behaviour may well be *desirable* — but it is unstated and unpinned: the amended §5.2 exclusion list names only *"an incarnation-owned offload panic"*, and the new doc comments at `raw.rs:1410` / `raw.rs:1453` say *"offload-work panic"*. Nothing pins it, so it can regress silently in either direction.

**Fix (pick one):** widen §5.2 and both doc comments to "any incarnation-owned disposal panic observed at a receive boundary" and add a pinning test; or tag the slot so only offload-work payloads resume at the receive boundary and freeze-time destructor panics stay on the cleanup path.

### 3. P3 nit — `SPEC.md:1001`

The edit leaves a 34-char orphan line (`` `on_stop`. Grace bounds drain plus ``) inside a paragraph hard-wrapped at ~70 columns.

### 4. Test quality — mutation-checked, all four are non-vacuous

| mutation | tests that go red |
|---|---|
| remove `raw.rs:1428` (local self-stop `recv`) | `pending_offload_panic_resumes_from_local_recv_before_discard_teardown` **only** |
| remove `raw.rs:1437` (externally-cancelled `recv`) | `externally_stopped_recv_resumes_a_pending_offload_panic` **only** |
| remove `raw.rs:1459` (stopping `try_recv`) | `stopping_try_recv_resumes_a_pending_offload_panic_before_drain` **only** |
| remove all three (= main) | those three **plus** `pending_offload_panic_stops_before_drain_on_stop_and_decorator_resume` |

Each production line has exactly one unique pin; nothing else across `offloads`/`actor`/`drain`/`raw` (62 tests) is disturbed. The added branch coverage in `a8a0618` does what its commit message claims.

One observation: the `Drain` decorator test is a *composite* — under `Drain`, `recv`'s local-stop branch and `try_recv` cover for each other, so it survives every single-line mutation. It pins the user-visible #263 contract but contributes no unique line coverage. Its sibling's comment ("Discard bypasses Handler's `try_recv` drain… pins the local-stop branch of `recv`") is accurate.

No `assert!` inside any handler — all four tests judge from the test body via `AtomicBool`/`Mutex` evidence or the published exit message. The two raw tests use `unreachable!()` inside `run`, but the discriminating assertion is the published exit *message* in the test body; that pin is mutation-proven to work today, though it leans on epilogue primary-over-cleanup precedence to surface the wrong message, so it is slightly indirect.

### 5. Cleared

- **Lock rule.** No new `.lock()` sites. `PanicSlot::take` (`raw.rs:254-259`) releases the guard inside the method, so `resume_preferred_panic_outside_unwind` (`raw.rs:1063-1066`) always resumes with no mutex held — including the unwind resumption itself starts. `PanicSlot::record` (`raw.rs:242-251`) computes `rejected` in a block and calls `discard_panic` after unlock.
- **Double resume / silent drop.** `take()` is atomic under the mutex, so exactly one of {receive-boundary resume, epilogue `take_resource_panic` at `2031`/`2037`, `RawResources::drop` at `1100`} wins a given payload. The only discard is `discard_panic(cleanup)` at `raw.rs:2050` — the pre-existing, documented primary-over-cleanup precedence.
- **Precedence.** `record` keeps the *first* payload, and `freeze()` cancels offloads before clearing continuations/timers/events, so a retained offload-work panic still outranks freeze-time destructor panics, and the actor's own panic still outranks both via `record_primary_panic`. Matches SPEC.

### 6. Commands

```
./tools/dev cargo nextest run -p shelterwood --test offloads     → 29/29 (baseline)
mutations A/B/C/D on raw.rs:1428/1437/1459, --no-fail-fast over
   offloads + actor + drain + raw                                → table above
temp repro 1 (in-flight offload panic, multi_thread)             → on_stop=true, decorator=Ok
temp repro 2 (freeze-time continuation destructor panic)         → PR vs main behaviour differs
./tools/dev cargo nextest run --workspace                        → 660/660 pass
```

All heavy runs under `nice -n 10 taskset -c 4-8`. All mutations and both repros reverted; worktree clean at `a8a0618`.

File: crates/shelterwood/src/raw.rs [hunk 1]
```diff
@@ -1433,6 +1434,7 @@ impl<M: Send + 'static> RawContext<M> {
                 // what makes the frozen accepted prefix a complete drain
                 // boundary once cancellation is observable here.
                 self.freeze_and_report();
+                self.resources.resume_pending_panic();
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply crates/shelterwood/src/raw.rs:1445]: **CONFIRMED**, and reproduced on this branch. The escape I looked for — "the offload is polled by the actor task, so the interleaving is unreachable" — does not exist: `raw.rs:1569` spawns the poller via `spawn_actor_work`, which is `tokio::spawn` (`shelterwood-runtime/src/spawn.rs:124-131`). The stop path only requests (`cancel` + `task.abort()`, `raw.rs:971-985`); the sole join is `join_offloads` (`raw.rs:1069-1087`), reached from the epilogue at `raw.rs:2035`, after `RawActor::run` has returned. Repro: `on_stop_ran = true`, `decorator_result = Some("ok")`, exit `Panicked` — #263's original symptom table.

Two corrections to the framing:

1. **Severity P2, not blocking.** Not a regression (main behaves identically here) and the published exit is still `Panicked`, so supervision/restart classification is unaffected. What the repro falsifies is the sentence this PR newly added to SPEC §5.2, not the fix itself.
2. **The prescribed remedy is not implementable as stated.** `try_recv` is synchronous and cannot join an offload task, and hoisting `join_offloads` ahead of the drain would change stop ordering for every actor. The correct resolution is to condition the guarantee on retention at the receive boundary, mirroring how §5.5 already handles `Guard::finished()` ("They are not a join guarantee"). Suggested wording in https://github.com/ralexstokes/shelterwood/pull/286#issuecomment-5332873880.

Separately, chasing this surfaced an unrelated P2: because `PanicSlot` is shared by every `RawDisposal::record`, the three new resume sites also promote freeze-time *user destructor* panics (discarded continuations, timers, queued offload events) from epilogue `cleanup` to incarnation-primary — a real behaviour change that neither the SPEC text nor the doc comments mention, and that nothing pins.

**Comment by ralexstokes** [comment]
Review findings resolved in `fa775db` + `53aaa35`.

**§1 — Codex P2 (`raw.rs:1437`), the SPEC sentence overclaimed.** Confirmed, and the prescribed remedy ("the stop transition must synchronize with in-progress polls") is not implementable at that site: `try_recv` is synchronous and cannot join an offload task, and hoisting `join_offloads` ahead of the drain would change stop ordering for every actor. So the guarantee is now stated **conditionally on retention**, mirroring how §5.5 already handles `Guard::finished()` ("They are not a join guarantee"):

> A disposal panic **already retained when a receive boundary is reached** therefore fails the incarnation before any further delivery or `on_stop`. As with `Guard::finished()` (§5.5) this is not a join: a panic that lands after the last receive boundary is still the incarnation's exit, but cannot suppress `on_stop`.

Why the conditional form is worth having rather than vacuous: the happens-before is real. `SharedOffloadFuture::poll` records into the disposal slot (`raw.rs:1560`) *before* firing `finished` (`raw.rs:949`), so an ordinary-completion `Guard::finished().await` — what all the new tests do — guarantees a populated slot at the next receive boundary. Cancellation fires the same latch (`raw.rs:921`) **without** recording, which is exactly the case the conditional excludes. §5.5's `Guard` bullet now states that ordering so §5.2's condition is checkable from the spec rather than from the source.

**§2 — the change widened to any disposal panic, not just offload panics.** Taken as a deliberate widening (option 1), not narrowed by tagging the slot. `PanicSlot` is shared by every `RawDisposal::record`, and `next_ready` has always resumed the same slot on the live path, so tagging would have split behaviour between the live and stopping boundaries for no contract gain. §5.2 and both doc comments now say "any incarnation-owned disposal panic observed at a receive boundary", and §5.2 defines the set (offload future + continuation closure, plus the queued continuations, armed timers and queued offload completions released at the intake freeze) and explicitly excludes the frozen mailbox prefix's own detached disposal, which stays a disposal fault per the paragraph above it.

Pinned by two new tests built on your repro shape (`continue_with` a message whose `Drop` panics, then `ctx.stop()`), one per `MailboxShutdown`:

| mutation | newly-red |
|---|---|
| remove `recv` local-stop resume (A) | `discarded_continuation_destructor_panic_resumes_before_discard_teardown` **uniquely** |
| remove `recv` external-stop resume (B) | none of the two |
| remove stopping `try_recv` resume (C) | none of the two |
| remove all three (= main) | both |

Under A alone the failure is exactly the reported table: `on_stop must not run after an incarnation-owned disposal panic` — i.e. `on_stop_ran = true`, `decorator_result = Some("ok")`, exit still `Panicked("discarded continuation destructor panic")`. The `Discard` variant is the unique pin for line A; the `Drain` variant is a composite (as with its offload sibling, `recv`'s local-stop branch and `try_recv` cover for each other) and pins the user-visible contract only. No `assert!` inside a handler — the actor publishes `AtomicBool`/`Mutex` evidence and the test body judges; the `Rejected`/delivered payloads are `mem::forget`ed so a regression reports through a flag rather than a second panic from inside the handler.

**§3 — orphan line at `SPEC.md:1001`.** Paragraph re-wrapped; the whole block is now 62–73 columns with no short line.

`./tools/dev just ci` green (exit 0), 662 tests.
